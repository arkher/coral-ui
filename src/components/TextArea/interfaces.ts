import {
  KeyboardTypeOptions,
  NativeSyntheticEvent,
  ReturnKeyTypeOptions,
  TextInputChangeEventData,
} from 'react-native';
import { InputFowardEvents } from '../Input/interfaces';

type CustomOnChange = NativeSyntheticEvent<TextInputChangeEventData> & {
  current: InputFowardEvents | null;
};

type TextAreaProps = {
  variant: 'small' | 'medium';
} & Partial<{
  label: string;
  placeholder: string;
  status: 'success' | 'error' | undefined;
  maxLength: number;
  assistiveText: string;
  keyboardType: KeyboardTypeOptions;
  autoCapitalize: 'none' | 'sentences' | 'words' | 'characters' | undefined;
  returnKeyType: ReturnKeyTypeOptions;
  value: string;
  onChange: ((e: CustomOnChange) => void) | undefined;
}>;

export { TextAreaProps };
