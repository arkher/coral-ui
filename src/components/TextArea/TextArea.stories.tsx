import React from 'react';
import { withDesign } from 'storybook-addon-designs';
import TextArea from './TextArea';

export default {
  title: 'Experimental/TextArea',
  component: TextArea,
  argTypes: {
    label: {
      control: { type: 'text' },
    },
    placeholder: {
      control: { type: 'text' },
    },
    status: {
      options: ['enable', 'success', 'error'],
      control: { type: 'inline-radio' },
    },
  },
  decorators: [withDesign],
};

interface TextInputProps {
  label: string;
  placeholder: string;
  // eslint-disable-next-line react/no-unused-prop-types
  status: 'success' | 'error' | undefined;
}

const urlHandoff =
  'https://www.figma.com/file/3raVfIADTUZCzFOOaQ9PMQ/HANDOFF-%7C-Core-Components-Mobile-%7C-Institucional?node-id=222%3A4012';

export const Small = ({
  label,
  placeholder,
  status,
}: TextInputProps): React.ReactNode => (
  <TextArea
    label={label}
    placeholder={placeholder}
    variant="small"
    status={status}
  />
);

Small.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Medium = ({
  label,
  placeholder,
}: TextInputProps): React.ReactNode => (
  <TextArea label={label} placeholder={placeholder} variant="medium" />
);

Medium.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};
