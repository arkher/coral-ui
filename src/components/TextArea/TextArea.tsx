import React, { useCallback, useRef, useState, useEffect } from 'react';
import { SafeAreaView, ScrollView } from 'react-native';
import { useTheme } from '@shopify/restyle';
import { Theme } from '../../themes/default';

import Text from '../Text';
import Input from '../Input';
import Box from '../Box';
import { TextAreaProps } from './interfaces';

const TextArea: React.FC<TextAreaProps> = ({
  label,
  placeholder,
  variant,
  status,
  maxLength,
}) => {
  const { colors, textVariants } = useTheme<Theme>();

  const [variantArea] = useState<CustomHeightComponent>(() => {
    switch (variant) {
      case 'small':
        return 'smallArea';
      case 'medium':
        return 'mediumArea';
      default:
        return variant;
    }
  });

  useEffect(() => {
    if (!variant) {
      throw new Error('Variant Ã© um campo definitivo.');
    }
  }, [variant]);

  return (
    <SafeAreaView>
      {!!label && (
        <Text fs="xxxs" fontWeight="700" color="neutralDarkest">
          {label}
        </Text>
      )}

      <Input
        placeholder={placeholder}
        placeholderTextColor={colors.neutralDark}
        variant={variantArea}
        multiline
        numberOfLines={7}
        my="quarck"
        p="xs"
      />

      {maxLength && (
        <Box flexDirection="row" justifyContent="flex-end">
          <Text color="neutralDark">000/{maxLength}</Text>
        </Box>
      )}
    </SafeAreaView>
  );
};

export default TextArea;
