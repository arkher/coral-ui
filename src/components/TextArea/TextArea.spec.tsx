import React from 'react';
import { cleanup, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import TextArea from './TextArea';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

test('should have a placeholder passed by prop', () => {
  const { getAllByPlaceholderText } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextArea variant="medium" placeholder="Placeholder" />
    </ThemeProvider>,
  );

  expect(getAllByPlaceholderText('Placeholder').length).toBe(1);
});

test('should TextArea render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextArea variant="medium" placeholder="Placeholder" />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
