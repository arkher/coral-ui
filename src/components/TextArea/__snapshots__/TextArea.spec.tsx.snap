// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should TextArea render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#5D6A75",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "marginVertical": 4,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Array [
          Object {
            "borderWidth": 1,
            "height": 56,
          },
        ],
      ]
    }
    testID="Box"
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      numberOfLines={7}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Placeholder"
      placeholderTextColor="#5D6A75"
      rejectResponderTermination={true}
      selectionColor="#5D6A75"
      style={
        Array [
          Object {},
          Object {
            "flex": 1,
            "fontFamily": "\\"Inter\\", sans-serif",
            "textAlignVertical": "top",
          },
        ]
      }
      testID="Input"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-end",
        },
        Array [
          Object {},
        ],
      ]
    }
    testID="Box"
  />
</RCTSafeAreaView>
`;
